pm: stm32: support U5x series Stop3 low-power mode

ABESTM opened the pull request:
Implements STOP3 low-power mode for STM32U5 devices. Uses similar approach as suspend to RAM on STM32WBA. LPTIM is not available in STOP3 mode, so RTC is used for STOP3 timing.
Also, should fix behavior of suspend to RAM on STM32WBA.

~~Waiting for #73983 to be merged~~

ABESTM committed:
soc: st: Add support for STOP3 on STM32U5

LPTIM is not available in STOP3 mode, so RTC needs to be used instead.
This code usese similar approach as STM32WBAx for suspend to ram.
The STOP3 is disabled by default in device tree.

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

ABESTM committed:
samples: boards: stm32: Add example for STOP3 mode on STM32U5

Simple blinky example, but using STOP3 mode. When in STOP3 mode,
GPIOs are not driven, but only pull-up or pull-down can be enabled
based on value in PWR registers.

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

erwango commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -84,18 +84,32 @@ static struct k_spinlock lock;
 #define cycle_t uint32_t
 
+typedef enum {
+	STM32_LPTIM_TIMEOUT_NONE = 0,
+#if defined(CONFIG_PM_S2RAM)
+	STM32_LPTIM_TIMEOUT_STANDBY = 1,
+#endif
+#if defined(CONFIG_STM32_USE_STOP3)
+	STM32_LPTIM_TIMEOUT_STOP3 = 2,
+#endif
```
You don't need to make both modes coexisting: S2RAM is only for WBA and WBA doesn't support STOP3

ABESTM replied:
In theory the U5 might support suspend to RAM, although the RAM is limited to SRAM2 and 64kB. So, I wanted to make the code more future proof.

erwango replied:
If this could allow to simplify the code, I'd prefer to not consider it for now. This would also fasten the review process.

ABESTM committed:
drivers: timer: Fix timing in suspend-to-ram

Fix timing in suspend-to-ram when using STM32WBA.
Switch to use RTC timer should be done only when idle is set
and LPTIM clocks should be switched off

Signed-off-by: Adam Berlinger <adam.berlinger@st.com>

erwango reviewed (changes requested):
Non functional review

erwango commented on soc/st/stm32/stm32u5x/power.c:
```diff
@@ -41,6 +70,31 @@ void set_mode_stop(uint8_t substate_id)
 		LL_PWR_SetPowerMode(LL_PWR_STOP2_MODE);
 		break;
+#ifdef CONFIG_STM32_USE_STOP3
+	case 4: /* enter STOP2 mode */
```
```suggestion
	case 4: /* enter STOP3 mode */
```

erwango commented on samples/boards/stm32/power_mgmt/stop3/boards/nucleo_u575zi_q.overlay:
```diff
@@ -0,0 +1,16 @@
+/ {
```
Missing copyright

erwango commented on samples/boards/stm32/power_mgmt/stop3/CMakeLists.txt:
Otherwise, you can add a README that specifies it

erwango commented on samples/boards/stm32/power_mgmt/stop3/src/main.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
```
```suggestion
 * Copyright (c) 2021 STMicroelectronics
```

erwango commented on samples/boards/stm32/power_mgmt/stop3/src/main.c:
```diff
@@ -0,0 +1,44 @@
+		 * So we enable pull-up/pull-down based on LED status
+		 */
+		if (led_is_on) {
+			PWR->PDCRB &= ~(uint32_t)(1 << 7);
+			PWR->PUCRB |= (1 << 7);
+		} else {
+			PWR->PUCRB &= ~(uint32_t)(1 << 7);
+			PWR->PDCRB |= (1 << 7);
```
Would be ideal to make it configurable from led.port and led.pin.
Doesn't have to be much complex, maybe using an array to get the address?
If too complex in short term, let's at minimum add a comment stating this is expecting such led value (or generating error is led value is not as expected)

ABESTM replied:
I used the code proposed by @ajarmouni-st and also compute the parameter from led.port.

erwango commented:
Otherwise LGTM

ajarmouni-st commented:
I think `STM32_USE_STOP3` is not clear, should be renamed to `STM32_PM_STOP3` or `STM32_STOP3_LP_MODE`....

ajarmouni-st commented on soc/st/stm32/stm32u5x/Kconfig.soc:
```diff
@@ -36,6 +38,12 @@ config SOC_STM32U5A9XX
+	bool
+	default $(dt_path_enabled,/cpus/power-states/state3) && $(dt_chosen_enabled,$(DT_CHOSEN_STDBY_TIMER))
+	help
+	  Allows STOP3 entry
```
```suggestion
	  Enable support for STM32 STOP3 low-power mode.
	  Based on the Cortex-M33 Deepsleep mode combined with peripheral clock gating.
	  All clocks in the core domain are stopped.
	  The PLL, MSIS, MSIK, HSI16, and HSE oscillators are disabled.
	  All SRAMs and register contents are preserved.
	  SRAMs can be totally or partially switched off to further reduce consumption.
```

ajarmouni-st commented on samples/boards/stm32/power_mgmt/stop3/src/main.c:
```diff
@@ -0,0 +1,44 @@
+		 * So we enable pull-up/pull-down based on LED status
+		 */
+		if (led_is_on) {
+			PWR->PDCRB &= ~(uint32_t)(1 << 7);
+			PWR->PUCRB |= (1 << 7);
```
```suggestion
			LL_PWR_DisableGPIOPullDown(LL_PWR_GPIO_PORTB, (1 << 7));
			LL_PWR_EnableGPIOPullUp(LL_PWR_GPIO_PORTB, (1 << 7));
```

ajarmouni-st commented on samples/boards/stm32/power_mgmt/stop3/src/main.c:
```diff
@@ -0,0 +1,44 @@
+			PWR->PDCRB &= ~(uint32_t)(1 << 7);
+			PWR->PUCRB |= (1 << 7);
+		} else {
+			PWR->PUCRB &= ~(uint32_t)(1 << 7);
+			PWR->PDCRB |= (1 << 7);
```
```suggestion
			LL_PWR_DisableGPIOPullUp(LL_PWR_GPIO_PORTB, (1 << led.pin));
			LL_PWR_EnableGPIOPullDown(LL_PWR_GPIO_PORTB, (1 << led.pin));
```

ajarmouni-st commented on samples/boards/stm32/power_mgmt/stop3/src/main.c:
```diff
@@ -0,0 +1,44 @@
+			PWR->PUCRB &= ~(uint32_t)(1 << 7);
+			PWR->PDCRB |= (1 << 7);
+		}
+		k_msleep(SLEEP_TIME_MS);
```
```suggestion
                /* activate PWR pull-ups/pull-downs */
                LL_PWR_EnablePUPDConfig();
                
		k_msleep(SLEEP_TIME_MS);
```

ABESTM replied:
Good point. Initially I called the ``LL_PWR_EnablePUPDConfig()`` in power.c. But I think this should be up to user if they want to use this.

ABESTM replied:
It can also be called once at the beginning of the example, since it takes effect only in STOP3 mode

ajarmouni-st replied:
Are you sure PWR pull-ups/downs have an effect in STOP3 mode? they only mention standby & shutdown here
https://github.com/STMicroelectronics/stm32u5xx_hal_driver/blob/5313c6919f88e8ec4212da91bbe1146d4262d44c/Inc/stm32u5xx_ll_pwr.h#L2650

ajarmouni-st replied:
Also shouldn't `LL_PWR_DisablePUPDConfig()` be called before reconfiguring PUPDs? (I'm just thinking out loud)

ABESTM commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -84,6 +84,10 @@ static struct k_spinlock lock;
 
+#if !defined(CONFIG_PM_S2RAM) && !defined(CONFIG_STM32_USE_STOP3)
+#error CONFIG_PM_S2RAM or CONFIG_STM32_USE_STOP3 needs to be selected
+#endif
```
I made the STM32_LPTIM_STDBY_TIMER non user selectable. I also updated the help description.

ajarmouni-st commented on soc/st/stm32/stm32u5x/Kconfig.soc:
```diff
@@ -5,6 +5,8 @@
 # Copyright (c) 2023 STMicroelectronics
 # SPDX-License-Identifier: Apache-2.0
 
+DT_CHOSEN_STDBY_TIMER := st,lptim-stdby-timer
```
Should be in `Kconfig` as well

erwango replied:
@ABESTM: last blocking comment

FRASTM commented on soc/st/stm32/stm32u5x/power.c:
```diff
@@ -26,6 +27,34 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+	/* Clear all wake-up flags */
+	uint32_t wusr = PWR->WUSR;
+
+	PWR->WUSCR = wusr;
```
```suggestion
	LL_PWR_ClearFlag_WU();
```

ABESTM replied:
Good point, I updated it

FRASTM commented on samples/boards/stm32/power_mgmt/stop3/README.rst:
```diff
@@ -0,0 +1,38 @@
+.. _stm32-pm-stop3:
+
+STM32 PM STOP3 mode
+#######################
```
```suggestion
###################
```
